problem-solver/cxx/exampleModule/searcher/PathSearcher.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+* This source file is part of an OSTIS project. For the latest info, see
+* http://ostis.net Distributed under the MIT License (See accompanying file
+* COPYING.MIT or copy at http://opensource.org/licenses/MIT)
+*/
+
+#include "PathSearcher.hpp"
+
+#include "keynodes/Keynodes.hpp"
+#include "utils/NumberUtils.hpp"
+
+PathSearcher::PathSearcher(ScMemoryContext * context)
+  : context(context)
+{
+}
+
+void PathSearcher::findPath(
+    ScAddr const & graph,
+    ScAddr const & startNode,
+    ScAddr const & endNode,
+    ScAddr const & connectorTemplateAddr,
+    ConnectorTemplateKeyElements const & connectorTemplateKeyElements,
+    ScAddr const & connectorWeightTemplateAddr,
+    WeightTemplateKeyElements const & weightTemplateKeyElements,
+    unsigned & pathLength,
+    ScAddrVector & path)
+{
+  ScAddrQueue vertexesToCheck;
+  ScAddrSet usedConnectors;
+  std::map<ScAddr, unsigned, ScAddrLessFunc> pathLengthToVertexes;
+  std::map<ScAddr, ScAddrVector, ScAddrLessFunc> pathsToVertexes;
+
+  vertexesToCheck.emplace(startNode);
+
+  while (!vertexesToCheck.empty())
+  {
+    ScAddr const & currnetVertex = vertexesToCheck.front();
+    vertexesToCheck.pop();
+
+    ScAddrToValueUnorderedMap<ScAddr> neighborsWithConnectors =
+        getNeighborsWithConnectors(graph, currnetVertex, connectorTemplateAddr, connectorTemplateKeyElements);
+
+    for (auto const & neighborWithConnector : neighborsWithConnectors)
+    {
+      ScAddr const & connector = neighborWithConnector.second;
+      if (usedConnectors.find(connector) != usedConnectors.cend())
+        continue;
+
+      ScAddr const & neighbor = neighborWithConnector.first;
+      unsigned const connectorWeight =
+          getConnectorWeights(connector, connectorWeightTemplateAddr, weightTemplateKeyElements);
+
+      if (pathLengthToVertexes.find(neighbor) == pathLengthToVertexes.cend())
+      {
+        pathLengthToVertexes[neighbor] = pathLengthToVertexes[currnetVertex] + connectorWeight;
+
+        pathsToVertexes[neighbor] = pathsToVertexes[currnetVertex];
+        pathsToVertexes[neighbor].emplace_back(neighbor);
+      }
+      else
+      {
+        unsigned newPathLength = pathLengthToVertexes[currnetVertex] + connectorWeight;
+        if (newPathLength < pathLengthToVertexes[neighbor])
+        {
+          pathLengthToVertexes[neighbor] = newPathLength;
+
+          pathsToVertexes[neighbor] = pathsToVertexes[currnetVertex];
+          pathsToVertexes[neighbor].emplace_back(neighbor);
+        }
```

Comment:
something with used connectors clearing should be here
![image](https://github.com/user-attachments/assets/5b4c0ff3-4664-41cb-9a7d-f4fae63fb09e)